clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -422,7 +434,17 @@ public boolean hasUnsent() {
                                         break;
                                     } else {
                                         RecordBatch batch = deque.pollFirst();
-                                        batch.close();
+                                        if (transactionState != null) {
+                                            log.debug("Dest: {} : pid: {}, epoch: {}, Assigning sequence for {}-{} : {}", node,
+                                                    transactionState.pidAndEpoch().pid, transactionState.pidAndEpoch().epoch,
+                                                    batch.topicPartition.topic(), batch.topicPartition.partition(),
+                                                    transactionState.sequenceNumber(batch.topicPartition));
+                                            batch.closeWithTransactionState(transactionState.pidAndEpoch().pid,
+                                                    transactionState.pidAndEpoch().epoch,
+                                                    transactionState.sequenceNumber(batch.topicPartition));
+                                        } else {
+                                            batch.close();
```

Comment:
See Jason's comment below. I added a `setProducerState` method, and removed the `closeWithTransactionState`. This way, we have only one close path, with the producer state being set only on a successful send if idempotence is enabled.